### PR TITLE
feat(metrics): Add HTTP RED metrics and remove gorilla/mux

### PR DESCRIPTION
**Pull request by kskewes-sf**
- update `/debug` landing page to correct /metrics URL
- add HTTP request duration and counter metrics per handler. Note durations aggregate all response codes to minimise cardinality
- remove `gorilla/mux` dependency to simplify route configuration and project has been seeking a maintainer for some time
- don't blindly prefix unhandled paths with Kubernetes current context value, e.g: `/not-a-handled-path` should return a 404. 

Note that some handlers return a 500 when it is more of a client error (missing query params).
Invalid/incorrect `webfiles` such as `/<context>/webfiles/not-a-real-file.css` also return 500's instead of 404. This is unchanged in this PR and something for later.

### CODE REVIEW

File: pkg/sloop/webserver/webserver.go [hunk 1]
```diff
@@ -66,23 +65,27 @@ type WebConfig struct {
...
+			Name: "sloop_webserver_http_requests_total",
+			Help: "A counter for requests to the wrapped handler.",
+		},
+		[]string{"code", "handler"},
+	)
+	metricWebServerRequestDuration = promauto.NewHistogramVec(
+		prometheus.HistogramOpts{
+			Name:    "sloop_webserver_http_request_duration_seconds",
+			Help:    "A histogram of latencies for requests to the wrapped handler.",
+			Buckets: []float64{.005, .01, .025, .05, .1, .25, .5, 1, 2.5, 5, 10, 30, 60},
```
Comment (kskewes-sf) [inline pkg/sloop/webserver/webserver.go:79]: Potentially the number of buckets could be reduced for now as we are querying a presumably local BadgerDB.
If the data store is changed in the future then it may be possible to see longer durations.
I could go either way.

**Comment by kskewes-sf** [comment]
Example metrics:
```
# make some requests
$ curl -s localhost:8080/metrics | grep http
# HELP sloop_webserver_http_request_duration_seconds A histogram of latencies for requests to the wrapped handler.
# TYPE sloop_webserver_http_request_duration_seconds histogram
sloop_webserver_http_request_duration_seconds_bucket{handler="debug",le="0.005"} 1
sloop_webserver_http_request_duration_seconds_bucket{handler="debug",le="0.01"} 1
sloop_webserver_http_request_duration_seconds_bucket{handler="debug",le="0.025"} 1
sloop_webserver_http_request_duration_seconds_bucket{handler="debug",le="0.05"} 1
sloop_webserver_http_request_duration_seconds_bucket{handler="debug",le="0.1"} 1
sloop_webserver_http_request_duration_seconds_bucket{handler="debug",le="0.25"} 1
sloop_webserver_http_request_duration_seconds_bucket{handler="debug",le="0.5"} 1
sloop_webserver_http_request_duration_seconds_bucket{handler="debug",le="1"} 1
sloop_webserver_http_request_duration_seconds_bucket{handler="debug",le="2.5"} 1
sloop_webserver_http_request_duration_seconds_bucket{handler="debug",le="5"} 1
sloop_webserver_http_request_duration_seconds_bucket{handler="debug",le="10"} 1
sloop_webserver_http_request_duration_seconds_bucket{handler="debug",le="30"} 1
sloop_webserver_http_request_duration_seconds_bucket{handler="debug",le="60"} 1
sloop_webserver_http_request_duration_seconds_bucket{handler="debug",le="+Inf"} 1
sloop_webserver_http_request_duration_seconds_sum{handler="debug"} 0.000429136
sloop_webserver_http_request_duration_seconds_count{handler="debug"} 1
sloop_webserver_http_request_duration_seconds_bucket{handler="index",le="0.005"} 1
sloop_webserver_http_request_duration_seconds_bucket{handler="index",le="0.01"} 1
sloop_webserver_http_request_duration_seconds_bucket{handler="index",le="0.025"} 1
sloop_webserver_http_request_duration_seconds_bucket{handler="index",le="0.05"} 1
sloop_webserver_http_request_duration_seconds_bucket{handler="index",le="0.1"} 1
sloop_webserver_http_request_duration_seconds_bucket{handler="index",le="0.25"} 1
sloop_webserver_http_request_duration_seconds_bucket{handler="index",le="0.5"} 1
sloop_webserver_http_request_duration_seconds_bucket{handler="index",le="1"} 1
sloop_webserver_http_request_duration_seconds_bucket{handler="index",le="2.5"} 1
sloop_webserver_http_request_duration_seconds_bucket{handler="index",le="5"} 1
sloop_webserver_http_request_duration_seconds_bucket{handler="index",le="10"} 1
sloop_webserver_http_request_duration_seconds_bucket{handler="index",le="30"} 1
sloop_webserver_http_request_duration_seconds_bucket{handler="index",le="60"} 1
sloop_webserver_http_request_duration_seconds_bucket{handler="index",le="+Inf"} 1
sloop_webserver_http_request_duration_seconds_sum{handler="index"} 0.000436966
sloop_webserver_http_request_duration_seconds_count{handler="index"} 1
sloop_webserver_http_request_duration_seconds_bucket{handler="root",le="0.005"} 3
sloop_webserver_http_request_duration_seconds_bucket{handler="root",le="0.01"} 3
sloop_webserver_http_request_duration_seconds_bucket{handler="root",le="0.025"} 3
sloop_webserver_http_request_duration_seconds_bucket{handler="root",le="0.05"} 3
sloop_webserver_http_request_duration_seconds_bucket{handler="root",le="0.1"} 3
sloop_webserver_http_request_duration_seconds_bucket{handler="root",le="0.25"} 3
sloop_webserver_http_request_duration_seconds_bucket{handler="root",le="0.5"} 3
sloop_webserver_http_request_duration_seconds_bucket{handler="root",le="1"} 3
sloop_webserver_http_request_duration_seconds_bucket{handler="root",le="2.5"} 3
sloop_webserver_http_request_duration_seconds_bucket{handler="root",le="5"} 3
sloop_webserver_http_request_duration_seconds_bucket{handler="root",le="10"} 3
sloop_webserver_http_request_duration_seconds_bucket{handler="root",le="30"} 3
sloop_webserver_http_request_duration_seconds_bucket{handler="root",le="60"} 3
sloop_webserver_http_request_duration_seconds_bucket{handler="root",le="+Inf"} 3
sloop_webserver_http_request_duration_seconds_sum{handler="root"} 0.0006570290000000001
sloop_webserver_http_request_duration_seconds_count{handler="root"} 3
sloop_webserver_http_request_duration_seconds_bucket{handler="webFile",le="0.005"} 2
sloop_webserver_http_request_duration_seconds_bucket{handler="webFile",le="0.01"} 2
sloop_webserver_http_request_duration_seconds_bucket{handler="webFile",le="0.025"} 2
sloop_webserver_http_request_duration_seconds_bucket{handler="webFile",le="0.05"} 2
sloop_webserver_http_request_duration_seconds_bucket{handler="webFile",le="0.1"} 2
sloop_webserver_http_request_duration_seconds_bucket{handler="webFile",le="0.25"} 2
sloop_webserver_http_request_duration_seconds_bucket{handler="webFile",le="0.5"} 2
sloop_webserver_http_request_duration_seconds_bucket{handler="webFile",le="1"} 2
sloop_webserver_http_request_duration_seconds_bucket{handler="webFile",le="2.5"} 2
sloop_webserver_http_request_duration_seconds_bucket{handler="webFile",le="5"} 2
sloop_webserver_http_request_duration_seconds_bucket{handler="webFile",le="10"} 2
sloop_webserver_http_request_duration_seconds_bucket{handler="webFile",le="30"} 2
sloop_webserver_http_request_duration_seconds_bucket{handler="webFile",le="60"} 2
sloop_webserver_http_request_duration_seconds_bucket{handler="webFile",le="+Inf"} 2
sloop_webserver_http_request_duration_seconds_sum{handler="webFile"} 0.005156629
sloop_webserver_http_request_duration_seconds_count{handler="webFile"} 2
# HELP sloop_webserver_http_requests_total A counter for requests to the wrapped handler.
# TYPE sloop_webserver_http_requests_total counter
sloop_webserver_http_requests_total{code="200",handler="debug"} 1
sloop_webserver_http_requests_total{code="200",handler="index"} 1
sloop_webserver_http_requests_total{code="200",handler="metrics"} 4
sloop_webserver_http_requests_total{code="200",handler="webFile"} 1
sloop_webserver_http_requests_total{code="307",handler="root"} 1
sloop_webserver_http_requests_total{code="404",handler="root"} 2
sloop_webserver_http_requests_total{code="500",handler="webFile"} 1
```

File: go.sum [hunk 2]
```diff
@@ -20,6 +20,7 @@ github.com/Azure/go-autorest/autorest/adal v0.5.0/go.mod h1:8Z9fGy2MpX0PvDjB1pEg
 github.com/Azure/go-autorest/autorest/date v0.1.0 h1:YGrhWfrgtFs84+h0o46rJrlmsZtyZRg470CqAXTZaGM=
 github.com/Azure/go-autorest/autorest/date v0.1.0/go.mod h1:plvfp3oPSKwf2DNjlBjWF/7vwR+cUD/ELuzDCXwHUVA=
 github.com/Azure/go-autorest/autorest/mocks v0.1.0/go.mod h1:OTyCOPRA2IgIlWxVYxBee2F5Gr4kF2zd2J5cFRaIDN0=
+github.com/Azure/go-autorest/autorest/mocks v0.2.0 h1:Ww5g4zThfD/6cLb4z6xxgeyDa7QDkizMkJKe0ysZXp0=
```
Comment (kskewes-sf) [inline go.sum:23]: `go mod tidy` had not been run on `master` so this was missing.

**Review by sarjamil — APPROVED** [review]

**Review by sarjamil — APPROVED** [review]

File: go.mod [hunk 3]
```diff
@@ -13,7 +13,6 @@ require (
 	github.com/golang/glog v0.0.0-20160126235308-23def4e6c14b
 	github.com/golang/protobuf v1.4.3
 	github.com/googleapis/gnostic v0.3.1 // indirect
-	github.com/gorilla/mux v1.8.0
```
[the remark this answers is not in the document]
Comment (kskewes-sf) [reply go.mod:16]: Yes and because we had a couple of different route handling patterns. This specifically made it hard to follow how routes were handled and what middleware applied to each route without looking up Gorill/mux docs.

[the remark this answers is not in the document]
Comment (kskewes-sf) [reply go.mod:16]: > This repository has been archived by the owner on Dec 10, 2022. It is now read-only.

The gorilla toolkit suite of repo's have just been archived.
